### PR TITLE
Make core.internal.atomic handle floats (dmd)

### PR DESCRIPTION
**Pull request by rikkimax**

### CODE REVIEW

**Comment by rikkimax** [comment]
@WalterBright ``atomicLoad`` in ``core.internal.atomic`` has been rewritten to be calling convention agnostic.

Needs a review before continuing as I'm far too rusty and this is compiler backend developer-owned.

It'll be good to get rid of some of the cruft inside of ``core.atomic`` that only exists because of it using inline assembly that is calling convention specific.

**Review by WalterBright — APPROVED** [review]

**Comment by RazvanN7** [comment]
@rikkimax Looks like @WalterBright has approved this PR. Is it ready to be merged? (it's currently still a draft)

**Comment by rikkimax** [comment]
> @rikkimax Looks like @WalterBright has approved this PR. Is it ready to be merged? (it's currently still a draft)

Not even close.

A quick count of modified functions, there are four more that need their inline assembly replaced.

I only pinged Walter to get a review because I wanted to be sure that what I'm doing with it would be ok. I'm too rusty with x86 assembly to not request such a review.

**Comment by rikkimax** [comment]
Ok, I was right. EBX/RBX are not scratch registers and must be reset.

Next push will have that fix.

File: druntime/src/core/stdc/stdatomic.d [hunk 1]
```diff
@@ -1116,9 +1192,48 @@ A atomic_fetch_and_explicit_impl(A, M)(shared(A)* obj, M arg, memory_order order
...
+            return atomicFetchSub!order(obj, arg);
+        }
+    }
+    else
+    {
+        // copied from core.atomic
+        static if (op == "+=" || op == "-="  || op == "*="  || op == "/=" ||
+                op == "%=" || op == "^^=" || op == "&="  || op == "|=" ||
+                op == "^=" || op == "<<=" || op == ">>=" || op == ">>>=") // skip "~="
+        {
```
Comment (ibuclaw) [inline druntime/src/core/stdc/stdatomic.d:1225]: This list seems unnecessary when you just need `+=` or `-=`

Comment (rikkimax) [reply druntime/src/core/stdc/stdatomic.d:1225]: Currently yes.

I need to cleanup other parts that do this operation manually.

**Comment by rikkimax** [comment]
Once Iain has had a chance to look at it again, and the CI is green, I'm happy with it.

File: druntime/src/core/stdc/stdatomic.d (hunk 1 above)
Comment (rikkimax) [reply druntime/src/core/stdc/stdatomic.d:1134]: Some of those are now used.

**Comment by RazvanN7** [comment]
cc @ibuclaw

**Comment by rikkimax** [comment]
@kinke I've included your changes here for asserting out on the bad fail memory orders.
@ibuclaw please review the ones for GDC, I can expand upon this for GDC if it is not ISA specific.

**Comment by kinke** [comment]
Thx, that was quick! :+1: - I'd need it in `stable` though; so if you don't want to extract it to its own PR, please extract it as separate commit at least, so that I can cherry-pick it for LDC v1.37. [And based on cppreference, I'd expect the assertion to be desirable for all compilers and targets, simplifying the implementation without extra `version` etc.]

**Comment by rikkimax** [comment]
I was not aware stdatomic was in stable.

I am more than happy to rebase into stable.

**Comment by ibuclaw** [comment]
The static asserts should be part of core.internal.atomic though. As stdatomic is just the dumb wrapper?

See https://github.com/dlang/dmd/pull/15970 for all invalid atomic operations not caught by core.atomic (but really should).

**Comment by rikkimax** [comment]
Okay @ibuclaw CI is green again.

**Review by ibuclaw — APPROVED** [review]

**Pull request merged**